File: Polygon.py
from __future__ import annotations

from copy import copy
from dataclasses import dataclass, field

from Point import Point

@dataclass
class Polygon:
    vertices: list = field(default_factory=list)

    def __iadd__(self, other: Vector) -> Polygon:
        for vertex in self.vertices:
            vertex += other
        return self
    def __add__(self, other: Vector) -> Polygon:
        new = copy(self)
        new += other
        return new

    def __isub__(self, other: Vector) -> Polygon:
        for vertex in self.vertices:
            vertex -= other
        return self
    def __sub__(self, other: Vector) -> Polygon:
        new = copy(self)
        new -= other
        return new

File: Polygon.ts
import { Point } from './Point.js';
import { Vector } from './Vector.js';

export class Polygon {
	vertices: Point[];
	constructor(list: Point[]) {
		this.vertices = list;
	}

	add(vector: Vector) {
		return new Polygon(this.vertices.map((vertex) => vertex.add(vector)));
	}

	subtract(vector: Vector) {
		return new Polygon(this.vertices.map((vertex) => vertex.subtract(vector)));
	}
}
